src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperCollector.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+public abstract partial class TagHelperCollector<T>
+    where T : ITagHelperDescriptorProvider
+{
+    // This type is generic to ensure that each descendent gets its own instance of this field.
+    private static readonly ConditionalWeakTable<IAssemblySymbol, Cache> s_perAssemblyCaches = new();
+
+    private readonly Compilation _compilation;
+    private readonly ISymbol? _targetSymbol;
+
+    protected TagHelperCollector(Compilation compilation, ISymbol? targetSymbol)
+    {
+        _compilation = compilation;
+        _targetSymbol = targetSymbol;
+    }
+
+    private static bool IsTagHelperAssembly(IAssemblySymbol assembly)
+    {
+        // This as a simple yet high-value optimization that excludes the vast majority of
+        // assemblies that (by definition) can't contain a component.
+        return assembly.Name != null && !assembly.Name.StartsWith("System.", StringComparison.Ordinal);
+    }
+
+    protected abstract void Collect(ISymbol symbol, ICollection<TagHelperDescriptor> results);
+
+    public void Collect(TagHelperDescriptorProviderContext context)
+    {
+        if (_targetSymbol is not null)
+        {
+            Collect(_targetSymbol, context.Results);
+        }
+        else
+        {
+            Collect(_compilation.Assembly.GlobalNamespace, context.Results);
+
+            foreach (var reference in _compilation.References)
+            {
+                if (_compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
+                {
+                    if (!IsTagHelperAssembly(assembly))
+                    {
+                        continue;
+                    }
+
+                    TagHelperDescriptor[]? tagHelpers;
+
+                    lock (s_perAssemblyCaches)
```

Comment:
Small nit: CWTs are already threadsafe. I get needing to lock when we do initial population of an individual cache, but I would love it if we could not lock unless the cache for this assembly hasn't been populated.

Reply:
To avoid the lock, I believe we'd need to:
1. Switch to `s_perAssemblyCaches.GetOrCreate` to create the empty cache.
2. Use `Interlocked.CompareExchange` to set the array location in `Cache.Add`.
3. Have `Cache.Add` return the final value (if it didn't win the race) so we can use the right set of helpers here.
4. Use `Volatile.Read` from `cache.TryGet` when reading the subarray location (probably not _strictly_ necessary, worst that happens is we do some extra calculation, but good for correctness nonetheless).

Reply:
Yeah, I didn't initially use `GetOrCreate` because it was just tracking `TagHelperDescriptor[]`. When I moved it over to use the `Cache` type, I now have a public constructor and can use `GetOrCreate`. Thanks for the suggestion!